File: arrays_hashing/1_two_sum.cpp
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> hash;
        for (int i=0; i < nums.size(); i++) {
            // first check if we've found the answer
            if (hash.find(target - nums[i]) != hash.end()) {
                vector<int> ans = {i, hash.at(target - nums[i])};
                return ans;
            }
            hash[nums[i]] = i;
        }
        return {};
    }
};

File: arrays_hashing/1_two_sum.py
class Solution(object):
    def twoSum(self, nums, target):
        # single pass hash approach, O(n) time and space
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        hash = {}
        for i in range(len(nums)):
            # first check if we've found the answer
            if (target - nums[i]) in hash:
                return (i, hash[target - nums[i]])
            hash[nums[i]] = i